### PR TITLE
Modifications to next hid_counter generation

### PR DESCRIPTION
**Pull request by jdavcs**
A few modifications triggered by migration to SQLAlchemy 2.0 (ref #12651 

1. Key issue 1: bound metadata is no longer supported in SQLAlchemy 2.0 ([ref](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#implicit-and-connectionless-execution-bound-metadata-removed)). The previous implementation relied on a bind accessible via the Table object's metadata property (see documentation: [order of resolution](https://docs.sqlalchemy.org/en/20/orm/contextual.html#sqlalchemy.orm.scoping.scoped_session.get_bind)). 
Solution: use an explicit connection.
2. Key issue 2: autocommit mode is no longer supported in SQLAlchemy 2.0 ([ref](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#autocommit-mode-removed-from-session-autobegin-support-added) and [ref](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#library-level-but-not-driver-level-autocommit-removed-from-both-core-and-orm)). Thus, for this to work, an explicit commit (or session flush) would be required. However, this operation should not affect the state of the session (as described in f63dc3b9df7a313c649d125f73d0ac6aa693514a).
Solution: use SQLAlchemy Core to obtain a connection from the engine: this will limit the commit to the operation.
3. Expire the hid_counter attribute. Once the operation commits, the value of the hid_counter attribute in the session is no longer consistent with the database. Expiring it will force it to be reloaded on the next access. This affects *only* this attribute (I've added a test verifying this behavior).
Note: this potentially affects current behavior whenever we access a history's hid_counter: (a) [in the `ModelImportStore`](https://github.com/galaxyproject/galaxy/blob/837332f6d1564c755b01225363b920444f34ec28/lib/galaxy/model/store/__init__.py#L148), (b) in the [`dev-detailed` and `betawebclient` views](https://github.com/galaxyproject/galaxy/blob/837332f6d1564c755b01225363b920444f34ec28/lib/galaxy/managers/histories.py#L353) and the [`_serialize` History method](https://github.com/galaxyproject/galaxy/blob/837332f6d1564c755b01225363b920444f34ec28/lib/galaxy/model/__init__.py#L2603), (c) in the [legacy history controller](https://github.com/galaxyproject/galaxy/blob/837332f6d1564c755b01225363b920444f34ec28/lib/galaxy/webapps/galaxy/controllers/history.py#L63), and (d) in the [`copy` History method](https://github.com/galaxyproject/galaxy/blob/837332f6d1564c755b01225363b920444f34ec28/lib/galaxy/model/__init__.py#L2582). The effects would be (1) a call to the database to refresh the attribute, and (2) the value may be different (but correct). But our tests pass.
4. Update SQLAlchemy sql expression syntax. Because some of it was 15 years old.
5. Move the whole thing into the model. There must have been a reason why it was initially placed into the module with all the mappings, but that was years ago, and with all the updates to the model and SQLAlchemy I don't think there's any reason to keep it there. Besides, the overridden code was never accessed.

Ping @assuntad23, @dannon: although (I think) this is unlikely, this might affect the client (iff there was a case when the supplied value of hid_counter was incorrect - i.e., outdated due to calls to db_next_hid)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/model/__init__.py [hunk 1]
```diff
@@ -2440,15 +2441,33 @@ def add_pending_items(self, set_output_hid=True):
...
+        Handle with SQLAlchemy Core to keep this independent from current session state, except:
+        expire hid_counter attribute, since its value in the session is no longer valid.
+        """
+        session = object_session(self)
+        engine = session.bind
+        table = self.__table__
+        history_id = cached_id(self)
+        update_stmt = update(table).where(table.c.id == history_id).values(hid_counter=table.c.hid_counter + n)
+
+        with engine.begin() as conn:
```
Comment (mvdbeek) [inline lib/galaxy/model/__init__.py:2459]: That creates a new commit, which we don't want. Why is this needed ?

Comment (jdavcs) [reply lib/galaxy/model/__init__.py:2459]: item 2 in the PR description. In my understanding, there's no autocommit in SA2.0. i.e., it would work now using `with endine.connect() as conn`, but unless we did `conn.commit()` in the end, it wouldn't work as soon as we moved to SA2.0.

Comment (mvdbeek) [reply lib/galaxy/model/__init__.py:2459]: Can you verify we're not expiring all attributes here ? If we are this would be a huge performance penalty. It seems we might not be, since you're explicitly expiring hid_counter ?

Comment (jdavcs) [reply lib/galaxy/model/__init__.py:2459]: I added a test for that very reason :smiley:  4ea118584da2bcc2e4be0080abdac56032308faa

File: test/unit/data/test_galaxy_mapping.py [hunk 2]
```diff
@@ -920,6 +920,31 @@ def test_can_manage_private_dataset(self):
...
+        self.persist(u, h)
+        state = inspect(h)
+        assert h.hid_counter == 1
+        assert 'hid_counter' not in state.unloaded
+        assert 'id' not in state.unloaded
+
+        h._next_hid()
+
+        assert 'hid_counter' in state.unloaded  # this attribute has been expired
+        assert 'id' not in state.unloaded  # but other attributes have NOT been expired
```
Comment (mvdbeek) [inline test/unit/data/test_galaxy_mapping.py:935]: Oh, so because we bypass the current session completely this works. That's neat!

**Review by mvdbeek — APPROVED** [review]

File: test/unit/data/test_galaxy_mapping.py (hunk 2 above)
Comment (mvdbeek) [reply test/unit/data/test_galaxy_mapping.py:935]: And if it didn't it would've probably wreaked havoc in the tests, where a lot of HDAs would be committed with a hid, which would raise exceptions now.

Comment (jdavcs) [reply test/unit/data/test_galaxy_mapping.py:935]: Yep! I also turned on echo=True for the engine when I was working on this to make sure there were no surprises.

**Comment by ic4f** [comment]
Thank you!
